src/EditorFeatures/Core/Implementation/CommentSelection/AbstractCommentSelectionBase.cs

```diff
@@ -0,0 +1,208 @@
+
+                var selectedSpans = textView.Selection.GetSnapshotSpansOnBuffer(subjectBuffer);
+                if (selectedSpans.IsEmpty())
+                {
+                    return true;
+                }
+
+                var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document == null)
+                {
+                    return true;
+                }
+
+                var service = document.GetLanguageService<ICommentSelectionService>();
+                if (service == null)
+                {
+                    return true;
+                }
+
+                var edits = CollectEditsAsync(document, service, subjectBuffer, selectedSpans, command, cancellationToken).WaitAndGetResult(cancellationToken);
+
+                ApplyEdits(document, textView, subjectBuffer, service, title, edits);
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Applies the requested edits and sets the selection.
+        /// This operation is not cancellable.
+        /// </summary>
+        private void ApplyEdits(Document document, ITextView textView, ITextBuffer subjectBuffer,
+            ICommentSelectionService service, string title, CommentSelectionResult edits)
+        {
+            // Create tracking spans to track the text changes.
+            var currentSnapshot = subjectBuffer.CurrentSnapshot;
+            var trackingSpans = edits.TrackingSpans
+                .SelectAsArray(textSpan => (originalSpan: textSpan, trackingSpan: CreateTrackingSpan(edits.ResultOperation, currentSnapshot, textSpan.TrackingTextSpan)));
+
+            // Apply the text changes.
+            using (var transaction = new CaretPreservingEditTransaction(title, textView, _undoHistoryRegistry, _editorOperationsFactoryService))
+            {
+                document.Project.Solution.Workspace.ApplyTextChanges(document.Id, edits.TextChanges.Distinct(), CancellationToken.None);
+                transaction.Complete();
+            }
+
+            // Convert the tracking spans into snapshot spans for formatting and selection.
+            var trackingSnapshotSpans = trackingSpans.Select(s => CreateSnapshotSpan(subjectBuffer.CurrentSnapshot, s.trackingSpan, s.originalSpan));
+
+            if (trackingSnapshotSpans.Any())
+            {
+                if (edits.ResultOperation == Operation.Uncomment)
+                {
+                    // Format the document only during uncomment operations.  Use second transaction so it can be undone.
+                    using (var transaction = new CaretPreservingEditTransaction(title, textView, _undoHistoryRegistry, _editorOperationsFactoryService))
+                    {
+                        var formattedDocument = Format(service, subjectBuffer.CurrentSnapshot, trackingSnapshotSpans, CancellationToken.None);
+                        if (formattedDocument == null)
```

Comment:
you don't need to explicitly cancel. using will do that for you if you don't complete